stdlib/README.adoc

```diff
@@ -0,0 +1,208 @@
+being of general interest in 2018, at different degrees, and might
+thus deserve some lightweight support in the stdlib: some (limited)
+support for Unicode (e.g. Uchar, perhaps utf8<->latin1 conversion),
+simple encoding/escaping stuff (e.g. Base64), a few more
+datastructures (e.g. bitsets), perhaps even a definition of a JSON
+datatype/parser/printer. Opinions will necessarily vary here. Who will
+decide, and based on which criteria?
+
+== Criteria for extending stdlib
+
+The size in itself should not really be a direct criterion for
+accepting or rejecting a proposed addition to the stdlib, I'd say. We
+should rather focus on other criteria (which are affected by the size,
+but not only):
+
+- The efforts required to maintain it (taking backward compatibility
+  into account).
+- The code and API quality (including the documentation).
+- How general purpose the addition is.
+- The usefulness of extending the stdlib vs maintaining the addition
+  as an external project.
+- In particular, the risk than an external project can do something
+  clearly better (which would leave us in the uncomfortable situation
+  where we need to keep something in the stdlib for backward
+  compatibility while knowing it isn't a good solution to recommend).
+
+For instance, adding a small function to an existing opaque
+datastructure (Set, Map, Stack, Buffer) would generally induce a very
+small maintenance cost and it's not really possible to do that outside
+the stdlib (short of completely reimplementing the datastructure
+altogether). The argument is weaker for extending e.g. List or Array,
+since this can more easily be done outside stdlib. That said, if the
+addition is useful enough, there is no strong reason not to accept it.
+
+
+== More to discuss
+
+* When a module `M` has a main type it must be named `M.t`.
+* Functions should take the main type `t` as first argument (unless
+  there is a good reason for it, see next point).
+* If a functions takes another function as argument, that argument
+  should have the label `f` and not be last (to be able to omit the
+  label).
+* When a function may raise an exception as a non-exceptional
+  behavior, its name must have the suffix `_exn` to remind the user
+  that the exception should be handled.  Another function returning an
+  option instead must be provided.  Example:
+[source,ocaml]
+----
+val find : 'a t -> key -> 'a option
+val find_exn : 'a t -> key -> 'a
+----
+
+
+=== Collections
+
+* If a module `M` defines a collection of values, say `'a M.t`, it
+  must also define the following standard higher order functions:
+
+[source,ocaml]
+----
+val iter : f:('a -> unit) -> 'a t -> unit
+val iteri : f:(int -> 'a -> unit) -> 'a t -> unit
+val map : f:('a -> 'b) -> 'a t -> 'b t
+val mapi : f:(int -> 'a -> 'b) -> 'a t -> 'b t
+val fold : f:('a -> 'b -> 'a) -> init:'a -> 'b t -> 'a
+----
+
+The `iter`, `map` and `fold` function need not to specify the order in
```

Comment:
I'm not sure to interpret that correctly, but it would be better, I think, to specifying the ordering when there is a natural one (e.g. increasing order for sets), and explicitly say "unspecified" otherwise.  Even if it would be nice to always keep the ordering undocumented to allow later changes to the internal representation, people will nevertheless depend on the ordering if it is nice enough.

Reply:
Not specifying the order for `map` and `iter` would allow for possible parallel execution. I don't think it makes sense not to specify the order of `fold` though.

Reply:
If Array.iter does not make the ordering explicit, one cannot use it for e.g.:

````ocaml
let dump f a =
  Array.iter (fun x -> Printf.printf " - %s\n" (f x)) a
````

In my opinion, this seriously reduces the usefulness for `iter`.  Are you suggesting that for those use case, one should use `fold` (with a specified ordering) and a `unit` accumulator?

Reply:
No I think you are right on `iter` (in fact I tend not to define `iter` on my data structures, using `fold` as you suggested if needed).

Reply:
One option is to have parallel execution variants of map and iter in which the ordering isn't specified and the implementation is free to use deep parallelism.

Reply:
Yes, I think the client code would need to be explicit anyway about parallelism (if only because of race conditions).  Anyway, this is mostly theoretical as long as multicore is not mainstream.

Reply:
(And the signature is likely to be slightly different for a parallel version,
*e.g.* to let the user specify the size of "chunks".)

Reply:
> I think, to specifying the ordering when there is a natural one (e.g. increasing order for sets), and explicitly say "unspecified" otherwise.

That is one possibility but one has to read the documentation (as the code is less explicit).  Another way would be to have `iter_left` or `iter_incr` to specify that one cares about the order.  In my opinion this is especially useful for `fold`: there are many commutative operators for which the order is not important.  One also wants some consistency with `to_seq` — there could be several of these transformations differing by specifying the order of the sequence.

Reply:
FTR, I think unspecified evaluation order is a billion-dollar mistake of the same order as null pointers.

If you want to make a parallel version, it's much better to be explicit about it.